feat: implement user deposit cap

MeLlamoPablo opened the pull request:
This PR implements the user deposit cap alongside with the global deposit cap. The mock contract defined in #885 is replaced with [this one](https://goerli.etherscan.io/address/0x92680aa94e19176ac5a2dd20a43573521d6b8949#code), which also implements `getUserDepositAmount` and `getUserDepositCap`.

### Scenario 1: user cap aleady exceeded

![Screenshot 2023-02-17 at 10 36 13](https://user-images.githubusercontent.com/11708035/219607983-7840b44c-ba4f-40d2-9c55-c04faf2021f3.png)

### Scenario 2: user cap would be exceeded by deposit

https://user-images.githubusercontent.com/11708035/219607056-e27a7a06-56b5-4d8f-a347-010f72b03156.mp4

# FE PR Checklist

- [x] Self-reviewed each line of code, preempted questions about logic or decisions
- [x] Noted any intentional deviations from design
- [x] Tried hard to break it and failed, think like a malicious user trying to find edge cases to break the app
- [x] Tested on supported device sizes
    - [x] iPhone 12 Pro
    - [x] iPad Air
    - [x] 1920x1080 (how to add custom size: [https://umaar.com/dev-tips/51-add-new-device/](https://umaar.com/dev-tips/51-add-new-device/))
- [x] Tested all widths between 1920px wide to 390px wide and confirmed no weird breakages at widths in between
- [x] (if related to balances) Tested with empty wallet, modest wallet, whale wallet
- [ ] (if browser specific logic) Tested on Brave, Firefox, Safari
- [x] (if styling) Tested Light/Dark mode

chrisling-dev reviewed (changes requested):
When showing "exceeded-if-deposit", wdyt if we show whichever limit that is smaller? Currently it's a bit weird that when I exceed global limit (100k) and click on it, it shows another limit $500, then I have to click the $500 again to be able to deposit.

MeLlamoPablo commented on apps/frontend/core-dapp/abi/depositRecord.abi.json:
```diff
@@ -25,6 +25,38 @@
     "stateMutability": "view",
     "type": "function"
   },
+  {
```
Sure! Where can I get that from?

chrisling-dev replied:
Oh you need to go to `prepo-monorepo/apps/smart-contracts/core`, run `yarn` which generates the artifacts, and find it in `artifacts/contracts/DepositRecord.sol/DepositRecord.json`:

```
[
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "contract IAccountList",
          "name": "accountList",
          "type": "address"
        }
      ],
      "name": "AccountListChange",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "address",
          "name": "hook",
          "type": "address"
        },
        {
          "indexed": false,
          "internalType": "bool",
          "name": "allowed",
          "type": "bool"
        }
      ],
      "name": "AllowedHooksChange",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "uint256",
          "name": "cap",
          "type": "uint256"
        }
      ],
      "name": "GlobalNetDepositCapChange",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": true,
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "indexed": true,
          "internalType": "bytes32",
          "name": "previousAdminRole",
          "type": "bytes32"
        },
        {
          "indexed": true,
          "internalType": "bytes32",
          "name": "newAdminRole",
          "type": "bytes32"
        }
      ],
      "name": "RoleAdminChanged",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "indexed": false,
          "internalType": "bytes32",
          "name": "newRoleAdminNominee",
          "type": "bytes32"
        }
      ],
      "name": "RoleAdminNomineeUpdate",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": true,
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "indexed": true,
          "internalType": "address",
          "name": "account",
          "type": "address"
        },
        {
          "indexed": true,
          "internalType": "address",
          "name": "sender",
          "type": "address"
        }
      ],
      "name": "RoleGranted",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "indexed": false,
          "internalType": "address",
          "name": "account",
          "type": "address"
        },
        {
          "indexed": false,
          "internalType": "bool",
          "name": "nominated",
          "type": "bool"
        }
      ],
      "name": "RoleNomineeUpdate",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": true,
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "indexed": true,
          "internalType": "address",
          "name": "account",
          "type": "address"
        },
        {
          "indexed": true,
          "internalType": "address",
          "name": "sender",
          "type": "address"
        }
      ],
      "name": "RoleRevoked",
      "type": "event"
    },
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "uint256",
          "name": "cap",
          "type": "uint256"
        }
      ],
      "name": "UserDepositCapChange",
      "type": "event"
    },
    {
      "inputs": [],
      "name": "DEFAULT_ADMIN_ROLE",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "SET_ACCOUNT_LIST_ROLE",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "SET_ALLOWED_HOOK_ROLE",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "SET_GLOBAL_NET_DEPOSIT_CAP_ROLE",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "SET_USER_DEPOSIT_CAP_ROLE",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        }
      ],
      "name": "acceptRole",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        }
      ],
      "name": "acceptRoleAdmin",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "getAccountList",
      "outputs": [
        {
          "internalType": "contract IAccountList",
          "name": "",
          "type": "address"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "getGlobalNetDepositAmount",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "",
          "type": "uint256"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "getGlobalNetDepositCap",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "",
          "type": "uint256"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        }
      ],
      "name": "getRoleAdmin",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        }
      ],
      "name": "getRoleAdminNominee",
      "outputs": [
        {
          "internalType": "bytes32",
          "name": "",
          "type": "bytes32"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "uint256",
          "name": "index",
          "type": "uint256"
        }
      ],
      "name": "getRoleMember",
      "outputs": [
        {
          "internalType": "address",
          "name": "",
          "type": "address"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        }
      ],
      "name": "getRoleMemberCount",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "",
          "type": "uint256"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "getUserDepositAmount",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "",
          "type": "uint256"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [],
      "name": "getUserDepositCap",
      "outputs": [
        {
          "internalType": "uint256",
          "name": "",
          "type": "uint256"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "grantRole",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "hasRole",
      "outputs": [
        {
          "internalType": "bool",
          "name": "",
          "type": "bool"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "hook",
          "type": "address"
        }
      ],
      "name": "isHookAllowed",
      "outputs": [
        {
          "internalType": "bool",
          "name": "",
          "type": "bool"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "isNominated",
      "outputs": [
        {
          "internalType": "bool",
          "name": "",
          "type": "bool"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "user",
          "type": "address"
        },
        {
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256"
        }
      ],
      "name": "recordDeposit",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256"
        }
      ],
      "name": "recordWithdrawal",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "renounceRole",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "revokeNomination",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "address",
          "name": "account",
          "type": "address"
        }
      ],
      "name": "revokeRole",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "contract IAccountList",
          "name": "accountList",
          "type": "address"
        }
      ],
      "name": "setAccountList",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "address",
          "name": "hook",
          "type": "address"
        },
        {
          "internalType": "bool",
          "name": "allowed",
          "type": "bool"
        }
      ],
      "name": "setAllowedHook",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "uint256",
          "name": "globalNetDepositCap",
          "type": "uint256"
        }
      ],
      "name": "setGlobalNetDepositCap",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes32",
          "name": "role",
          "type": "bytes32"
        },
        {
          "internalType": "bytes32",
          "name": "roleAdminNominee",
          "type": "bytes32"
        }
      ],
      "name": "setRoleAdminNominee",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "uint256",
          "name": "userDepositCap",
          "type": "uint256"
        }
      ],
      "name": "setUserDepositCap",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
    {
      "inputs": [
        {
          "internalType": "bytes4",
          "name": "interfaceId",
          "type": "bytes4"
        }
      ],
      "name": "supportsInterface",
      "outputs": [
        {
          "internalType": "bool",
          "name": "",
          "type": "bool"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    }
  ]
```

MeLlamoPablo replied:
Cool! Yeah, we should use the proper ABI, and if my mock is implemented correctly, this change should be transparent.

Btw, I changed it to `IDepositRecord`, not `DepositRecord`. I believe that's all the frontend should need.

MeLlamoPablo commented:
Those are great suggestions! I've updated the code

> When showing "exceeded-if-deposit", wdyt if we show whichever limit that is smaller? Currently it's a bit weird that when I exceed global limit (100k) and click on it, it shows another limit $500, then I have to click the $500 again to be able to deposit.

Done. I've implemented a small "priority system" on `DepositWarning`, I think that was the most readable way I could come up with. The logic is now:

* If the global cap has been already exceeded, show that warning.
* If the user cap has been already exceeded, show that warning.
* If depositing would exceed any of the two caps, show the warning for the cap with the least remaining amount. (Even if depositing would exceed both caps).
* Otherwise, don't show warnings

> Thoughts on making a function called "getDepositLimit", which accepts any string/number value, and returns an object:
> 
> ```
>   state: 'exceeded-if-deposit' | 'already-exceeded' | ...,
>   amount?: ...
> ```
> 
> So the UI could potentially be simpler, and it's just a matter of some JS check to return the right value in mobx store?
> 
> Also that the limit check and etc can be reused by open trade page when deposit + trade is implementated

[Refactored to this](https://github.com/prepo-io/prepo-monorepo/blob/a9392de633f2234040261a2efcfada60cbc8ff73/apps/frontend/core-dapp/src/features/deposit/DepositStore.ts#L166-L207), what do you think?

I think it's nice because it simplified `DepositStore` and the `DepositWarning` component. Now there aren't a lot of properties scattered around. Every property is grouped nicely into `globalDepositLimitInfo` and `userDepositLimitInfo`

chrisling-dev commented on apps/frontend/core-dapp/src/features/deposit/DepositStore.ts:
```diff
@@ -113,44 +127,82 @@ export class DepositStore {
+    })
+  }
+
+  private getDepositLimitInfo({
+    amount,
+    cap,
+  }: {
```
We should be able to simplify this further, it should just take an input number, without the cap.

With the input number, we can just check
- if something isnt ready, return { status: 'loading' }
- if globalCapExceeded return { status: 'global-exceeded', capInUsd, amountInUsd }
- if userCapExceeded return { status: 'user-exceeded', capInUsd, amountInUsd }
- if inputWouldExceedEitherGlobalOrUser return { status: 'exceeded-if-deposit, remainingAmountInUsd: Math.min(globalRemainingAmount, userRemainingAmount)
- return 'not-exceeded'

Then reusing this would be straightforward, we can just do
```
get depositLimitInfo() {
    return this. getDepositLimitInfo(depositAmount)
 }
```

reusing the logic in deposit + trade later
```
get depositTradeLimitInfo() {
    return this. getDepositLimitInfo(tradeAmount)
 }
```

And UI code would be as simple as:
```
switch(depositLimt.status) {
  case 'global-exceeded':
    return <GlobalExceededUI />
  case 'user-exceeded':
    return <UserExceededUI />
  case 'exceeded':
    return "Deposit limit exceeded. Only xxx remaining" UI
  default:
    return null
}
```

Combining everything for max reusability, the usage would only take 2 lines here and 2 lines in open + trade page later:

```
  const { depositLimitInfo } = depositStore

  ... 
  return <>
     <DepositWarning depositLimit={depositLimitInfo} />
  <>
```

Let me know if this makes sense or am I missing something here

chrisling-dev replied:
This way in depositDisabled check we don't have to check 2 separate caps (global and user), and can just check the deposit limit info object

Also I think global + user cap should be computed together and never separated because they are directly relevant

MeLlamoPablo replied:
Mm the reason I did this separate was so that I could reuse the same logic for the withdraw cap... Let me think about how to make your suggestion and the withdraw cap compatible.

chrisling-dev replied:
ohh gotcha, what about combing both userDepositLimitInfo and globalDepositLimitInfo to achieve returning one single object rather than relying on 2 separate computed values?

chrisling-dev replied:
btw 'already-exceeded' for both global and user cap has the same UI so it can just return the same name if we do the single object approach

There's a slight UI mismatch in this PR in that, it has different copy for global and personal deposit limit when it has already exceeded

MeLlamoPablo replied:
I see that the copies are different in this PR? ([user](https://github.com/prepo-io/prepo-monorepo/pull/886/files#diff-dfc6a0930a7dc3ae01d96be39baca961d758215af8cbbec640a2b8f5582ad60dR94-R106) and [global](https://github.com/prepo-io/prepo-monorepo/pull/886/files#diff-dfc6a0930a7dc3ae01d96be39baca961d758215af8cbbec640a2b8f5582ad60dR122-R132))

chrisling-dev replied:
> I see that the copies are different in this PR? ([user](https://github.com/prepo-io/prepo-monorepo/pull/886/files#diff-dfc6a0930a7dc3ae01d96be39baca961d758215af8cbbec640a2b8f5582ad60dR94-R106) and [global](https://github.com/prepo-io/prepo-monorepo/pull/886/files#diff-dfc6a0930a7dc3ae01d96be39baca961d758215af8cbbec640a2b8f5582ad60dR122-R132))

They should be the same copy for both user and global cap

![Screenshot 2023-02-21 at 4 38 45 PM](https://user-images.githubusercontent.com/81092286/220292053-d73d2742-2b91-4dbf-a62e-778c959c4d0c.png)

MeLlamoPablo replied:
I've given this another pass, and I think it looks much better thanks to your suggestions :)

Here's what I've done:

* I've extracted the logic that computes individual limits into a [reusable pure function](https://github.com/prepo-io/prepo-monorepo/blob/636db71f2e45e0ae863d9435d286f21746dd4400/apps/frontend/core-dapp/src/utils/balance-limits.ts#L14-L55)
  * It could be reused by the withdraw store to calculate the withdraw cap
* I've kept the [`globalDepositLimitInfo` and `userDepositLimitInfo`](https://github.com/prepo-io/prepo-monorepo/blob/636db71f2e45e0ae863d9435d286f21746dd4400/apps/frontend/core-dapp/src/features/deposit/DepositStore.ts#L148-L186) computed values in `DepositStore`, which now bridge data to this utility function. But now they are private and not accessible by the UI.
* I've added a [`depositLimit` computed value](https://github.com/prepo-io/prepo-monorepo/blob/636db71f2e45e0ae863d9435d286f21746dd4400/apps/frontend/core-dapp/src/features/deposit/DepositStore.ts#L188-L228) that combines the global and user limits into one and decides which one to show following the logic you outlined.

    Basically, the logic that was done in the UI is now done in the store.
* Now `DepositWarning` consists of [a dumb switch that just displays data](https://github.com/prepo-io/prepo-monorepo/blob/636db71f2e45e0ae863d9435d286f21746dd4400/apps/frontend/core-dapp/src/features/deposit/DepositWarning.tsx#L58-L102).

Let me know what you think!

chrisling-dev commented on apps/frontend/core-dapp/src/features/deposit/DepositWarning.tsx:
```diff
@@ -52,55 +53,92 @@ const InlineTextButton = styled.button`
   }
 
-  return null
+  const warnings: Array<{ element: ReactElement; priority: number }> = []
```
I think the priority system is an overkill and somewhat hard to read, would it be easier if there's only one object that gives 'global-cap-exceeded', 'user-cap-exceeded', 'exceed-if-deposit'

And we just do a switch here?

```
switch(depositLimit.status) {
    case 'global-cap-exceeded':
    ...
}
```

chrisling-dev reviewed (approved):
nice work! LGTM
